sensor/mcp9808/mcp9808.go

```diff
@@ -0,0 +1,484 @@
+		BelowLower:    !(temp&(1<<13) == 0)}
+	tempResult.CelsiusDeg = wholeNum + (float64(fraction) / 10000.0)
+	if !(temp&(1<<12) == 0) { // read sign bit
+		tempResult.CelsiusDeg *= -1
+	}
+	return tempResult, nil
+}
+
+// readTempC reads from the ambient temperature register and returns the current temperature value in celsius
+func (d *MCP9808) readTempC(reg byte) (float64, error) {
+	temp, err := d.Bus.ReadWordFromReg(address, reg)
+	if err != nil {
+		return 0, err
+	}
+
+	log.Printf("\nTA vs TCrit: %v\nTA vs TUpper: %v\nTA vs TLower: %v\n", !(temp&(1<<15) == 0), !(temp&(1<<14) == 0), !(temp&(1<<13) == 0))
+
+	wholeTempC := temp &^ 0xF000
+
+	wholeNum := float64(wholeTempC&^0xF) / 16.0
+	fraction := wholeTempC &^ 0xFF0
+
+	tempRead := wholeNum + (float64(fraction) / 10000.0)
+	if !(temp&(1<<12) == 0) { // read sign bit
+		tempRead *= -1
+	}
+	return tempRead, nil
+}
+
+func (d *MCP9808) setTemp(reg byte, newTemp float64) error {
+	var signBit uint16
+	if newTemp < 0 {
+		newTemp *= -1
+		signBit = 0x1000
+	}
+	newTempWord := uint16(newTemp)*15 + uint16(newTemp*16) + signBit
+
+	if err := d.Bus.WriteWordToReg(address, reg, newTempWord); err != nil {
+		return err
+	}
+	return nil
+}
+
+// CriticalTempUpper reads the current temperature set in the critical temperature register.
+func (d *MCP9808) CriticalTempUpper() (float64, error) {
+	return d.readTempC(regCriticalTemp)
+}
+
+// SetCriticalTemp when the temperature goes above the set value the alert will be
+// triggered if enabled.
+func (d *MCP9808) SetCriticalTemp(newTemp float64) error {
+	d.cmu.Lock()
+	defer d.cmu.Unlock()
+
+	d.SetCriticalTempLock(false)
+	if _, err := d.WriteConfig(); err != nil {
+		return err
+	}
+
+	if err := d.setTemp(regCriticalTemp, newTemp); err != nil {
+		return err
+	}
+
+	d.SetCriticalTempLock(true)
+	_, err := d.WriteConfig()
+	return err
+}
+
+// WindowTempUpper reads the current temperature set in the upper window temperature register.
+func (d *MCP9808) WindowTempUpper() (float64, error) {
+	return d.readTempC(regUpperTemp)
+}
+
+// SetWindowTempUpper when the temperature goes above the set value the alert will be
+// triggered if enabled.
+func (d *MCP9808) SetWindowTempUpper(newTemp float64) error {
+	d.cmu.Lock()
+	defer d.cmu.Unlock()
+
+	d.SetWindowTempLock(false)
+	if _, err := d.WriteConfig(); err != nil {
+		return err
+	}
+
+	if err := d.setTemp(regUpperTemp, newTemp); err != nil {
+		return err
+	}
+
+	d.SetWindowTempLock(true)
+	_, err := d.WriteConfig()
+	return err
+}
+
+// WindowTempLower reads the current temperature set in the lower window temperature register.
+func (d *MCP9808) WindowTempLower() (float64, error) {
+	return d.readTempC(regLowerTemp)
+}
+
+// SetWindowTempLower when the temperature goes below the set value the alert will be
+// triggered if enabled.
+func (d *MCP9808) SetWindowTempLower(newTemp float64) error {
+	d.cmu.Lock()
+	defer d.cmu.Unlock()
+
+	d.SetWindowTempLock(false)
+	if _, err := d.WriteConfig(); err != nil {
```

Comment:
This is going to deadlock unless I'm missing something 'cause WriteConfig tries to acquire the lock you already hold. You need a writeConfig that doesn't lock that you can call here.